service/src/upgrade.rs

```diff
@@ -43,46 +68,255 @@ impl TryFrom<&String> for FailoverPolicy {
+        match cmd_map.get_mut(&cmd.mountpoint) {
+            Some(_) => {
+                cmd_map.remove(&cmd.mountpoint);
+                Ok(())
+            }
+            None => Err(Error::NotFound),
+        }
     }
 
     /// Disable online upgrade capability.
-    pub fn disable_upgrade(&mut self) {}
+    pub fn disable_upgrade(&mut self) {
+        self.disabled.store(true, Ordering::Release);
+    }
+
+    /// Save the fuse fds and fuse state data for online upgrade.
+    fn save(&mut self, data: &Vec<u8>) -> Result<()> {
+        self.backend
+            .save(&self.fds, data)
+            .map_err(UpgradeMgrError::StorageBackendError)?;
+        Ok(())
+    }
+
+    /// Restore the fuse fds and fuse state data for online upgrade.
+    fn restore(&mut self) -> Result<Vec<u8>> {
+        let mut fds = vec![0 as RawFd; 8];
+        let mut state_data = vec![0u8; MAX_STATE_DATA_LENGTH];
+        let (state_data_len, fds_len) = self
+            .backend
+            .restore(&mut fds, &mut state_data)
+            .map_err(UpgradeMgrError::StorageBackendError)?;
+        fds.truncate(fds_len);
+        state_data.truncate(state_data_len);
+
+        self.fds = fds;
+        Ok(state_data)
+    }
+
+    fn add_fd(&mut self, fd: RawFd) {
+        self.fds.push(fd);
+    }
 }
 
 /// Online upgrade utilities for FUSE daemon.
 pub mod fusedev_upgrade {
+    use std::fs::File;
+    use std::os::fd::{FromRawFd, RawFd};
+    use std::os::unix::io::AsRawFd;
+    use std::sync::atomic::Ordering;
+
+    use nydus_upgrade::persist::Snapshotter;
+    use versionize::{VersionMap, Versionize, VersionizeResult};
+    use versionize_derive::Versionize;
+
     use super::*;
-    use crate::fusedev::FusedevDaemon;
+    use crate::daemon::NydusDaemon;
+    use crate::fusedev::{FusedevDaemon, FusedevFsService};
+
+    #[derive(Versionize, Clone, Default)]
+    struct FusedevState {
+        backend_fs_mount_cmd_list: Vec<(FsBackendMountCmd, u8)>,
+        vfs_state_data: Vec<u8>,
+        fuse_conn_id: u64,
+    }
+
+    impl Snapshotter for FusedevState {
+        fn get_versions() -> Vec<HashMap<TypeId, u16>> {
+            vec![
+                // version 1
+                HashMap::from([(FusedevState::type_id(), 1)]),
+                // more versions for the future
+            ]
+        }
+    }
 
     /// Save state information for a FUSE daemon.
-    pub fn save(_daemon: &FusedevDaemon) -> Result<()> {
+    pub fn save(daemon: &FusedevDaemon) -> Result<()> {
+        let svc = daemon.get_default_fs_service().ok_or(Error::NotFound)?;
+
+        if !svc.get_vfs().initialized() {
+            return Err(Error::NotReady);
+        }
+
+        let mut mgr = svc.upgrade_mgr().unwrap();
+
+        // set fd
+        let fd = svc
+            .as_ref()
+            .as_any()
+            .downcast_ref::<FusedevFsService>()
+            .unwrap()
+            .session
+            .lock()
+            .unwrap()
+            .get_fuse_file()
```

Comment:
If nydusd exit when there are inflight fuse requests, IO hung occurs because the mainline kernel does not support resending inflight fuse requests. If we want to support takeover for fusedev, how about store the fuse fd created by ioctl FUSE_DEV_IOC_CLONE, then when nydusd exit, the inflight requests will be dropped instead of blocking IO.

Reply:
I'm not sure if macos can support clone,  maybe we need to keep the API more generic. 
Any suggestions? @imeoer @jiangliu

Reply:
Maybe make it as a built-in feature in `fuse-backend-rs`? WDYT cc @jiangliu ?

Reply:
Sure, let's get fuse-backend-rs ready for this:)

Reply:
@imeoer @jiangliu 

> Sure, let's get fuse-backend-rs ready for this:)

@imeoer @jiangliu I might be able to help with this :)

Reply:
https://github.com/cloud-hypervisor/fuse-backend-rs/pull/177 may fix this